src/tike/ptycho/solvers/lstsq.py

```diff
@@ -277,23 +283,67 @@ def _update_wavefront(
     varying_probe: npt.NDArray[cp.csingle],
     scan: npt.NDArray[cp.single],
     psi: npt.NDArray[cp.csingle],
+    exitwave_options: ExitWaveOptions,
     op: tike.operators.Ptycho,
 ) -> typing.Tuple[npt.NDArray[cp.csingle], float]:
 
     farplane = op.fwd(probe=varying_probe, scan=scan, psi=psi)
+
     intensity = cp.sum(
         cp.square(cp.abs(farplane)),
         axis=list(range(1, farplane.ndim - 2)),
     )
     costs = getattr(tike.operators,
                     f'{op.propagation.model}_each_pattern')(data, intensity)
+
     cost = cp.mean(costs)
-    logger.info('%10s cost is %+12.5e', 'farplane', cost)
-    farplane = -op.propagation.grad(data, farplane, intensity)
+
+    if exitwave_options.noise_model == 'poisson':
+
+        xi = 1 - data / (intensity + 1e-9)
+        grad_cost = farplane * xi[:, None, None, :, :]
+        abs2_Psi = cp.square(cp.abs(cp.swapaxes(cp.squeeze(farplane), 0,
+                                                1)))
+
+        Nspos = farplane.shape[0]
+        Nscpm = farplane.shape[2]
+
+        step_length = exitwave_options.step_length_start * cp.ones(
+            (Nscpm, Nspos))
+
+        if exitwave_options.step_length_usemodes == 'dominant_mode':
+
+            step_length = tike.ptycho.exitwave.poisson_steplength_ptychoshelves(
+                xi, intensity, data, exitwave_options.measured_pixels, step_length,
+                exitwave_options.step_length_weight)
+
+        else:
+
+            step_length = tike.ptycho.exitwave.poisson_steplength_approx(
+                xi, abs2_Psi, intensity, data, exitwave_options.measured_pixels,
+                step_length, exitwave_options.step_length_weight)
+
+        step_length = cp.swapaxes(step_length, 0, 1)[:, None, :, None, None]
+
+        chi = ((farplane - step_length * grad_cost) *
+               exitwave_options.measured_pixels +
+               farplane * exitwave_options.unmeasured_pixels_scaling *
+               exitwave_options.unmeasured_pixels)
+
+    else:
+
+        chi = (farplane * (cp.sqrt(data) /
+                               (cp.sqrt(intensity) + 1e-9))[..., None, None, :, :] *
+               exitwave_options.measured_pixels +
+               farplane * exitwave_options.unmeasured_pixels_scaling *
+               exitwave_options.unmeasured_pixels)
```

Comment:
Doesn't this update function push the intensity of the wavefront at unmeasured pixels towards zero? The measured pixels are updated according to the gradient direction with the computed step length. The unmeasured pixels are always decreased because `unmeasured_pixels_scaling` is `>=0,<1`?

Reply:
> The overall goal here would be to try to get the scaling of the unmeasured regions to smoothly transition to the scaling of the measured regions.

The unmeasured regions are unmeasured, so we can't know if they should smoothly transition or not.

In your example pdf, /gamma = 1 is equivalent to unmeasured_pixels_scaling = 1 which is equivalent to no constraint on the unmeasured regions. As long as that is the default behavior, I think it's fine for now.

Any non-1.0 value for unmeasured_pixels_scaling is enforcing a constraint on the updated wavefront because the gradient at that unmeasured pixel becomes non-zero.

Reply:
see new commit

Reply:
can I close this part of the conversation?